zellij-server/src/panes/plugin_pane.rs

```diff
@@ -139,127 +213,86 @@ impl Pane for PluginPane {
-            Ok(Some((vec![], Some(vte_output), vec![]))) // TODO: PluginPanes should have their own grid so that we can return the non-serialized TerminalCharacters and have them participate in the render buffer
-        } else {
-            Ok(None)
         }
+        Ok(None)
     }
     fn render_frame(
         &mut self,
-        _client_id: ClientId,
+        client_id: ClientId,
         frame_params: FrameParams,
         input_mode: InputMode,
     ) -> Result<Option<(Vec<CharacterChunk>, Option<String>)>> {
-        // FIXME: This is a hack that assumes all fixed-size panes are borderless. This
-        // will eventually need fixing!
-        let res = if self.frame && !(self.geom.rows.is_fixed() || self.geom.cols.is_fixed()) {
+        if self.borderless {
+            return Ok(None);
+        }
+        if let Some(grid) = self.grids.get(&client_id) {
+            let err_context = || format!("failed to render frame for client {client_id}");
             let pane_title = if self.pane_name.is_empty()
                 && input_mode == InputMode::RenamePane
                 && frame_params.is_main_client
             {
                 String::from("Enter name...")
             } else if self.pane_name.is_empty() {
-                self.pane_title.clone()
+                grid.title
+                    .clone()
+                    .unwrap_or_else(|| self.pane_title.clone())
             } else {
                 self.pane_name.clone()
             };
+
             let frame = PaneFrame::new(
                 self.current_geom().into(),
-                (0, 0), // scroll position
+                grid.scrollback_position_and_length(),
                 pane_title,
                 frame_params,
             );
-            Some(
-                frame
-                    .render()
-                    .with_context(|| format!("failed to render frame for client {_client_id}"))?,
-            )
+
+            let res = match self.frame.get(&client_id) {
+                // TODO: use and_then or something?
+                Some(last_frame) => {
+                    if &frame != last_frame {
+                        if !self.borderless {
+                            let frame_output = frame.render().with_context(err_context)?;
+                            self.frame.insert(client_id, frame);
+                            Some(frame_output)
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                },
+                None => {
+                    if !self.borderless {
+                        let frame_output = frame.render().with_context(err_context)?;
+                        self.frame.insert(client_id, frame);
+                        Some(frame_output)
+                    } else {
+                        None
+                    }
+                },
+            };
+            Ok(res)
```

Comment:
I've had a little fun with this one, hope you don't mind. This should be equivalent:

```rust
            self.frame
                .get(&client_id)
                .with_context(|| format!("failed to acquire last frame for client {}", client_id))
                .map_or_else(
                    |_err| !self.borderless,
                    |last_frame| (&frame != last_frame) && !self.borderless,
                )
                .then(|| {
                    let frame_output = frame.render().with_context(err_context)?;
                    self.frame.insert(client_id, frame);
                    Ok(frame_output)
                })
                .transpose()

```

Reply:
cool! Haven't looked at this specifically, but let's do it in a different PR. This bit was copy/pasted from elsewhere in the code - so we will have to change it there too (in terminal_pane.rs iirc)

Reply:
Also, on a side note... are you 100% sure this does the same thing? From a cursory read it seems not to, but I haven't gone deeply into this so maybe I'm wrong.